{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66aac959",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7fec9ce7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/mubaraq/Desktop/gnn-challenge/organizer_scripts'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e21287a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    Gene_ID    P1_1154    P1_1651    P1_1689     P1_1748  \\\n",
      "0     ENSG00000139055$ERP27  28.783956  19.929550  24.891419   39.407868   \n",
      "1      ENSG00000182759$MAFA  18.820279  26.223093  45.337942   33.185573   \n",
      "2       ENSG00000205221$VIT  23.248580  29.369864  13.334689  145.186882   \n",
      "3    ENSG00000207582$MIR30B  14.391978  14.684932  12.445710    2.074098   \n",
      "4  ENSG00000207620$MIR516A2  17.713204   7.342466   8.889793    7.259344   \n",
      "\n",
      "    P1_1973    P1_2013     P1_2052    P1_2139    P1_2263  ...    P9_1805  \\\n",
      "0   7.12230  44.643687   49.938054   6.726263  27.046152  ...  14.759850   \n",
      "1  10.68345  11.420478   16.646018  10.569842  36.782766  ...  28.465425   \n",
      "2  17.80575  53.987714  267.525291  73.988894  51.928611  ...  10.542750   \n",
      "3  11.87050  11.420478    1.189001   9.608947  12.982153  ...   9.488475   \n",
      "4   0.00000   8.305802    5.945006   7.687158   4.327384  ...   7.379925   \n",
      "\n",
      "        P931       P942       P950        P972       P976       P979  \\\n",
      "0   5.535609  13.722226  23.352076    8.807878   2.825287   8.498681   \n",
      "1  18.821069  27.444452  21.128069    2.935959  17.893483  16.390314   \n",
      "2  21.035313  72.270389  10.008033  137.990085   9.417622   5.463438   \n",
      "3  18.821069  24.700006  17.792058   10.765184  18.835245  13.355071   \n",
      "4  14.392582  30.188897  10.008033   53.825920  25.427580  10.926876   \n",
      "\n",
      "        P980       P990       P996  \n",
      "0  15.241932  10.832387  17.478337  \n",
      "1  44.773175  34.466687  11.958862  \n",
      "2   0.000000   4.923812   4.599562  \n",
      "3  16.194553   8.862862  12.878774  \n",
      "4  21.910277  10.832387   4.599562  \n",
      "\n",
      "[5 rows x 124 columns]\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------\n",
    "# 1. Load expression data\n",
    "# -----------------------------\n",
    "# Use relative path - notebook is in organizer_scripts/, data is in ../data/\n",
    "data_dir = '../data'\n",
    "\n",
    "expr_cfRNA = pd.read_csv(os.path.join(data_dir, 'expr_df_2_GSE192902.csv'))\n",
    "expr_placenta = pd.read_csv(os.path.join(data_dir, 'expr_df_GSE234729.csv'))\n",
    "\n",
    "meta_cfRNA = pd.read_csv(os.path.join(data_dir, 'metadata_cfRNA.csv'))\n",
    "meta_placenta = pd.read_csv(os.path.join(data_dir, 'metadata_placenta.csv'))\n",
    "\n",
    "\n",
    "print(expr_placenta.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e5d1ca62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared genes: 6650\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------\n",
    "# 2. Gene harmonization\n",
    "# -----------------------------\n",
    "\n",
    "\n",
    "# manipulating GENE numbers to ensure matching\n",
    "\n",
    "# Extract ENSG ID from 'ENSG...$SYMBOL'\n",
    "expr_placenta[\"ensembl_id\"] = expr_placenta[\"Gene_ID\"].str.split(\"$\").str[0]\n",
    "expr_placenta = expr_placenta.set_index(\"ensembl_id\")\n",
    "expr_placenta = expr_placenta.drop(columns=[\"Gene_ID\"])\n",
    "\n",
    "expr_cfRNA = expr_cfRNA.set_index(\"gene_num\")\n",
    "expr_cfRNA = expr_cfRNA.drop(columns=[\"gene_name\"])\n",
    "\n",
    "expr_placenta.index = expr_placenta.index.str.strip()\n",
    "expr_cfRNA.index = expr_cfRNA.index.str.strip()\n",
    "\n",
    "shared_genes = expr_placenta.index.intersection(expr_cfRNA.index)\n",
    "print(\"Shared genes:\", len(shared_genes))\n",
    "\n",
    "expr_placenta = expr_placenta.loc[shared_genes]\n",
    "expr_cfRNA = expr_cfRNA.loc[shared_genes]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3b3ae98",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# -----------------------------\n",
    "# 3. Normalize expression per gene (samples x genes format)\n",
    "\n",
    "'''\n",
    "We make samples the nodes and genes the features, scaling each gene across samples to produce meaningful....\n",
    "node feature vectors; the original orientation (genes as rows, samples as columns) would treat genes as ....\n",
    "“data points,” which doesn’t fit the GNN node-feature paradigm. because all the features comprises of its gene representation =)\n",
    "'''\n",
    "\n",
    "# -----------------------------\n",
    "# Transpose to get samples x genes, scale each gene across samples, then keep that format\n",
    "scaler = StandardScaler()\n",
    "expr_cfRNA_T = expr_cfRNA.T\n",
    "expr_cfRNA_scaled = pd.DataFrame(scaler.fit_transform(expr_cfRNA_T),\n",
    "                                  index=expr_cfRNA_T.index, columns=expr_cfRNA_T.columns)\n",
    "\n",
    "expr_placenta_T = expr_placenta.T\n",
    "expr_placenta_scaled = pd.DataFrame(scaler.transform(expr_placenta_T),\n",
    "                                     index=expr_placenta_T.index, columns=expr_placenta_T.columns)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2cd116f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(          ENSG00000169877  ENSG00000176463  ENSG00000205639  ENSG00000244716  \\\n",
       " 782752_3         0.156980        -0.072228         0.514708        -0.401727   \n",
       " 782752_4         0.444778         0.091410        -0.014561         0.108674   \n",
       " 637549_3         0.300879         0.255048         1.440930         0.108674   \n",
       " 637549_4        -0.130817        -0.399504        -0.279196        -0.656928   \n",
       " 549101_1        -0.130817        -0.276775        -0.345355        -0.912128   \n",
       " \n",
       "           ENSG00000124813  ENSG00000215302  ENSG00000226608  ENSG00000248309  \\\n",
       " 782752_3         0.144883         0.052908        -0.667910         0.868102   \n",
       " 782752_4        -0.033763        -0.172249         0.661549         0.036856   \n",
       " 637549_3        -0.033763         2.104340         0.661549         3.306422   \n",
       " 637549_4         0.055560         0.052908        -1.000274        -0.129393   \n",
       " 549101_1        -0.212409         0.778414        -0.667910         0.369354   \n",
       " \n",
       "           ENSG00000113356  ENSG00000226360  ...  ENSG00000278272  \\\n",
       " 782752_3         1.491133         0.146461  ...         0.520635   \n",
       " 782752_4         0.006385        -0.659073  ...         0.754390   \n",
       " 637549_3         3.295383        -0.659073  ...         0.520635   \n",
       " 637549_4        -0.024939        -1.061840  ...        -0.297506   \n",
       " 549101_1         0.394800        -0.659073  ...        -0.297506   \n",
       " \n",
       "           ENSG00000278311  ENSG00000278540  ENSG00000278588  ENSG00000278705  \\\n",
       " 782752_3         0.597109         0.238333         0.610908        -0.927578   \n",
       " 782752_4         0.109054        -0.134508        -0.206242         0.852804   \n",
       " 637549_3         0.426290         0.312902         0.610908        -0.136297   \n",
       " 637549_4        -0.159377        -0.059940        -0.001955        -0.334117   \n",
       " 549101_1        -0.061766        -0.059940        -0.819105        -0.334117   \n",
       " \n",
       "           ENSG00000278730  ENSG00000278771  ENSG00000278828  ENSG00000279117  \\\n",
       " 782752_3         1.134118        -0.046383         1.022826         0.392123   \n",
       " 782752_4         0.049723        -0.529131        -0.088537         0.134136   \n",
       " 637549_3         1.405217         0.214562         2.078621         0.478118   \n",
       " 637549_4        -0.266559        -1.031451        -0.174415         0.478118   \n",
       " 549101_1        -0.085826        -0.848789         0.305492        -0.123850   \n",
       " \n",
       "           ENSG00000281649  \n",
       " 782752_3         0.236402  \n",
       " 782752_4         0.146569  \n",
       " 637549_3         0.011820  \n",
       " 637549_4        -0.212762  \n",
       " 549101_1        -0.033096  \n",
       " \n",
       " [5 rows x 6650 columns],\n",
       "   Sample_title Sample_geo_accession          Sample_status  \\\n",
       " 0     944838_2           GSM5768318  Public on Jan 03 2022   \n",
       " 1     944838_1           GSM5768319  Public on Jan 03 2022   \n",
       " 2     944838_4           GSM5768320  Public on Jan 03 2022   \n",
       " 3     931746_1           GSM5768321  Public on Jan 03 2022   \n",
       " 4     826701_4           GSM5768322  Public on Jan 03 2022   \n",
       " \n",
       "   Sample_submission_date Sample_last_update_date Sample_type  \\\n",
       " 0            Jan 02 2022             Jan 03 2022         SRA   \n",
       " 1            Jan 02 2022             Jan 03 2022         SRA   \n",
       " 2            Jan 02 2022             Jan 03 2022         SRA   \n",
       " 3            Jan 02 2022             Jan 03 2022         SRA   \n",
       " 4            Jan 02 2022             Jan 03 2022         SRA   \n",
       " \n",
       "    Sample_channel_count Sample_source_name_ch1 Sample_organism_ch1   disease  \\\n",
       " 0                     1           blood plasma        Homo sapiens   control   \n",
       " 1                     1           blood plasma        Homo sapiens   control   \n",
       " 2                     1           blood plasma        Homo sapiens   control   \n",
       " 3                     1           blood plasma        Homo sapiens   control   \n",
       " 4                     1           blood plasma        Homo sapiens   control   \n",
       " \n",
       "    ... Sample_contact_zip/postal_code Sample_contact_country  \\\n",
       " 0  ...                          94305                    USA   \n",
       " 1  ...                          94305                    USA   \n",
       " 2  ...                          94305                    USA   \n",
       " 3  ...                          94305                    USA   \n",
       " 4  ...                          94305                    USA   \n",
       " \n",
       "   Sample_data_row_count Sample_instrument_model Sample_library_selection  \\\n",
       " 0                     0   Illumina NovaSeq 6000                     cDNA   \n",
       " 1                     0   Illumina NovaSeq 6000                     cDNA   \n",
       " 2                     0   Illumina NovaSeq 6000                     cDNA   \n",
       " 3                     0   Illumina NovaSeq 6000                     cDNA   \n",
       " 4                     0   Illumina NovaSeq 6000                     cDNA   \n",
       " \n",
       "    Sample_library_source Sample_library_strategy  \\\n",
       " 0         transcriptomic                 RNA-Seq   \n",
       " 1         transcriptomic                 RNA-Seq   \n",
       " 2         transcriptomic                 RNA-Seq   \n",
       " 3         transcriptomic                 RNA-Seq   \n",
       " 4         transcriptomic                 RNA-Seq   \n",
       " \n",
       "                                      Sample_relation  \\\n",
       " 0  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       " 1  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       " 2  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       " 3  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       " 4  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       " \n",
       "   Sample_supplementary_file_1 Series_sample_id  \n",
       " 0                        NONE       GSM5768318  \n",
       " 1                        NONE       GSM5768319  \n",
       " 2                        NONE       GSM5768320  \n",
       " 3                        NONE       GSM5768321  \n",
       " 4                        NONE       GSM5768322  \n",
       " \n",
       " [5 rows x 36 columns])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_cfRNA_scaled.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6f69e6c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>Sample_status</th>\n",
       "      <th>Sample_submission_date</th>\n",
       "      <th>Sample_last_update_date</th>\n",
       "      <th>Sample_type</th>\n",
       "      <th>Sample_channel_count</th>\n",
       "      <th>Sample_source_name_ch1</th>\n",
       "      <th>Sample_organism_ch1</th>\n",
       "      <th>disease</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_contact_zip/postal_code</th>\n",
       "      <th>Sample_contact_country</th>\n",
       "      <th>Sample_data_row_count</th>\n",
       "      <th>Sample_instrument_model</th>\n",
       "      <th>Sample_library_selection</th>\n",
       "      <th>Sample_library_source</th>\n",
       "      <th>Sample_library_strategy</th>\n",
       "      <th>Sample_relation</th>\n",
       "      <th>Sample_supplementary_file_1</th>\n",
       "      <th>Series_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>944838_2</td>\n",
       "      <td>GSM5768318</td>\n",
       "      <td>Public on Jan 03 2022</td>\n",
       "      <td>Jan 02 2022</td>\n",
       "      <td>Jan 03 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>blood plasma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>94305</td>\n",
       "      <td>USA</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>GSM5768318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>944838_1</td>\n",
       "      <td>GSM5768319</td>\n",
       "      <td>Public on Jan 03 2022</td>\n",
       "      <td>Jan 02 2022</td>\n",
       "      <td>Jan 03 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>blood plasma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>94305</td>\n",
       "      <td>USA</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>GSM5768319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>944838_4</td>\n",
       "      <td>GSM5768320</td>\n",
       "      <td>Public on Jan 03 2022</td>\n",
       "      <td>Jan 02 2022</td>\n",
       "      <td>Jan 03 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>blood plasma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>94305</td>\n",
       "      <td>USA</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>GSM5768320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>931746_1</td>\n",
       "      <td>GSM5768321</td>\n",
       "      <td>Public on Jan 03 2022</td>\n",
       "      <td>Jan 02 2022</td>\n",
       "      <td>Jan 03 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>blood plasma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>94305</td>\n",
       "      <td>USA</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>GSM5768321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>826701_4</td>\n",
       "      <td>GSM5768322</td>\n",
       "      <td>Public on Jan 03 2022</td>\n",
       "      <td>Jan 02 2022</td>\n",
       "      <td>Jan 03 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>blood plasma</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>94305</td>\n",
       "      <td>USA</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>GSM5768322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_title Sample_geo_accession          Sample_status  \\\n",
       "0     944838_2           GSM5768318  Public on Jan 03 2022   \n",
       "1     944838_1           GSM5768319  Public on Jan 03 2022   \n",
       "2     944838_4           GSM5768320  Public on Jan 03 2022   \n",
       "3     931746_1           GSM5768321  Public on Jan 03 2022   \n",
       "4     826701_4           GSM5768322  Public on Jan 03 2022   \n",
       "\n",
       "  Sample_submission_date Sample_last_update_date Sample_type  \\\n",
       "0            Jan 02 2022             Jan 03 2022         SRA   \n",
       "1            Jan 02 2022             Jan 03 2022         SRA   \n",
       "2            Jan 02 2022             Jan 03 2022         SRA   \n",
       "3            Jan 02 2022             Jan 03 2022         SRA   \n",
       "4            Jan 02 2022             Jan 03 2022         SRA   \n",
       "\n",
       "   Sample_channel_count Sample_source_name_ch1 Sample_organism_ch1   disease  \\\n",
       "0                     1           blood plasma        Homo sapiens   control   \n",
       "1                     1           blood plasma        Homo sapiens   control   \n",
       "2                     1           blood plasma        Homo sapiens   control   \n",
       "3                     1           blood plasma        Homo sapiens   control   \n",
       "4                     1           blood plasma        Homo sapiens   control   \n",
       "\n",
       "   ... Sample_contact_zip/postal_code Sample_contact_country  \\\n",
       "0  ...                          94305                    USA   \n",
       "1  ...                          94305                    USA   \n",
       "2  ...                          94305                    USA   \n",
       "3  ...                          94305                    USA   \n",
       "4  ...                          94305                    USA   \n",
       "\n",
       "  Sample_data_row_count Sample_instrument_model Sample_library_selection  \\\n",
       "0                     0   Illumina NovaSeq 6000                     cDNA   \n",
       "1                     0   Illumina NovaSeq 6000                     cDNA   \n",
       "2                     0   Illumina NovaSeq 6000                     cDNA   \n",
       "3                     0   Illumina NovaSeq 6000                     cDNA   \n",
       "4                     0   Illumina NovaSeq 6000                     cDNA   \n",
       "\n",
       "   Sample_library_source Sample_library_strategy  \\\n",
       "0         transcriptomic                 RNA-Seq   \n",
       "1         transcriptomic                 RNA-Seq   \n",
       "2         transcriptomic                 RNA-Seq   \n",
       "3         transcriptomic                 RNA-Seq   \n",
       "4         transcriptomic                 RNA-Seq   \n",
       "\n",
       "                                     Sample_relation  \\\n",
       "0  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       "1  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       "2  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       "3  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       "4  SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX...   \n",
       "\n",
       "  Sample_supplementary_file_1 Series_sample_id  \n",
       "0                        NONE       GSM5768318  \n",
       "1                        NONE       GSM5768319  \n",
       "2                        NONE       GSM5768320  \n",
       "3                        NONE       GSM5768321  \n",
       "4                        NONE       GSM5768322  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_cfRNA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f0c8ce37",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'diagnosis'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/gnn-challenge/venv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3812\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3811\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m3812\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3813\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/index.pyx:167\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/index.pyx:196\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7096\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: 'diagnosis'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# -----------------------------\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;66;03m# 4. Create train.csv (cfRNA) and test.csv (placenta)\u001b[39;00m\n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# -----------------------------\u001b[39;00m\n\u001b[32m      4\u001b[39m train_df = expr_cfRNA_scaled.copy()\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m train_df[\u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m] = \u001b[43mmeta_cfRNA\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mdiagnosis\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m  \u001b[38;5;66;03m# 1=preeclampsia, 0=control\u001b[39;00m\n\u001b[32m      6\u001b[39m train_df[\u001b[33m'\u001b[39m\u001b[33mnode_id\u001b[39m\u001b[33m'\u001b[39m] = [\u001b[33m'\u001b[39m\u001b[33mcfRNA_\u001b[39m\u001b[33m'\u001b[39m+\u001b[38;5;28mstr\u001b[39m(i) \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(train_df))]\n\u001b[32m      7\u001b[39m train_df = train_df.reset_index(drop=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/gnn-challenge/venv/lib/python3.13/site-packages/pandas/core/frame.py:4113\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4111\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.nlevels > \u001b[32m1\u001b[39m:\n\u001b[32m   4112\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._getitem_multilevel(key)\n\u001b[32m-> \u001b[39m\u001b[32m4113\u001b[39m indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4114\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[32m   4115\u001b[39m     indexer = [indexer]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/gnn-challenge/venv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3819\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3814\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   3815\u001b[39m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc.Iterable)\n\u001b[32m   3816\u001b[39m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[32m   3817\u001b[39m     ):\n\u001b[32m   3818\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[32m-> \u001b[39m\u001b[32m3819\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   3820\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[32m   3821\u001b[39m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[32m   3822\u001b[39m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[32m   3823\u001b[39m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[32m   3824\u001b[39m     \u001b[38;5;28mself\u001b[39m._check_indexing_error(key)\n",
      "\u001b[31mKeyError\u001b[39m: 'diagnosis'"
     ]
    }
   ],
   "source": [
    "\n",
    "# -----------------------------\n",
    "# 4. Create train.csv (cfRNA) and test.csv (placenta)\n",
    "# -----------------------------\n",
    "train_df = expr_cfRNA_scaled.copy()\n",
    "train_df['target'] = meta_cfRNA['diagnosis']  # 1=preeclampsia, 0=control\n",
    "train_df['node_id'] = ['cfRNA_'+str(i) for i in range(len(train_df))]\n",
    "train_df = train_df.reset_index(drop=True)\n",
    "train_df.to_csv(os.path.join(data_dir, 'train.csv'), index=False)\n",
    "\n",
    "test_df = expr_placenta_scaled.copy()\n",
    "test_df['node_id'] = ['placenta_'+str(i) for i in range(len(test_df))]\n",
    "test_df = test_df.reset_index(drop=True)\n",
    "test_df.to_csv(os.path.join(data_dir, 'test.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a58a92f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------\n",
    "# 4. Create train.csv (cfRNA) and test.csv (placenta) with correct label mapping\n",
    "# -----------------------------\n",
    "\n",
    "# Copy scaled expression matrices\n",
    "train_df = expr_cfRNA_scaled.copy()\n",
    "test_df = expr_placenta_scaled.copy()\n",
    "\n",
    "# Build a mapping from metadata sample ID -> numeric disease label\n",
    "# Adjust column name if your metadata uses 'disease' instead of 'diagnosis'\n",
    "disease_map = meta_cfRNA.set_index('Sample_title')['disease'].map({'control': 0, 'preeclampsia': 1})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72772053",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cfRNA dataset 209 valid samples with matching metadata\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Assign target labels by matching sample IDs (index of expr_cfRNA_scaled)\n",
    "train_df['target'] = train_df.index.map(disease_map)\n",
    "\n",
    "print(f'cfRNA dataset has {len(train_df['target'])} valid samples with matching metadata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e136ab2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "⚠️ Some cfRNA samples could not be matched to metadata!\n",
      "✅ train.csv and test.csv created with correct labels and node IDs\n"
     ]
    }
   ],
   "source": [
    "# Check for any unmatched samples\n",
    "if train_df['target'].isnull().any():\n",
    "    print(\"⚠️ Some cfRNA samples could not be matched to metadata!\")\n",
    "\n",
    "# Add unique node IDs\n",
    "train_df['node_id'] = ['cfRNA_'+str(i) for i in range(len(train_df))]\n",
    "test_df['node_id'] = ['placenta_'+str(i) for i in range(len(test_df))]\n",
    "\n",
    "# Reset indices (optional)\n",
    "train_df = train_df.reset_index(drop=True)\n",
    "test_df = test_df.reset_index(drop=True)\n",
    "\n",
    "# Save CSVs\n",
    "train_df.to_csv(os.path.join(data_dir, 'train.csv'), index=False)\n",
    "test_df.to_csv(os.path.join(data_dir, 'test.csv'), index=False)\n",
    "\n",
    "print(\"✅ train.csv and test.csv created with correct labels and node IDs\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "057747b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------\n",
    "# 5. Node types\n",
    "# -----------------------------\n",
    "node_types = pd.DataFrame({\n",
    "    'node_id': list(train_df['node_id']) + list(test_df['node_id']),\n",
    "    'node_type': ['cfRNA']*len(train_df) + ['placenta']*len(test_df)\n",
    "})\n",
    "node_types.to_csv(os.path.join(data_dir, 'node_types.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1701079",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# -----------------------------\n",
    "# 6. Build graph_edges.csv\n",
    "# -----------------------------\n",
    "edges = []\n",
    "\n",
    "# a) Within-modality cosine similarity edges (sparse)\n",
    "def build_edges(df, node_prefix, top_k=10):\n",
    "    sim_matrix = cosine_similarity(df[shared_genes])\n",
    "    for i in range(sim_matrix.shape[0]):\n",
    "        top_idx = np.argsort(sim_matrix[i])[-(top_k+1):-1]  # skip self\n",
    "        for j in top_idx:\n",
    "            edges.append({\n",
    "                'src': f'{node_prefix}_{i}',\n",
    "                'dst': f'{node_prefix}_{j}',\n",
    "                'edge_type': 'similarity'\n",
    "            })\n",
    "\n",
    "build_edges(train_df, 'cfRNA')\n",
    "build_edges(test_df, 'placenta')\n",
    "\n",
    "# b) Cross-modality edges (cfRNA <-> placenta)\n",
    "sim_matrix_cross = cosine_similarity(train_df[shared_genes], test_df[shared_genes])\n",
    "top_k = 5\n",
    "for i in range(sim_matrix_cross.shape[0]):\n",
    "    top_idx = np.argsort(sim_matrix_cross[i])[-top_k:]\n",
    "    for j in top_idx:\n",
    "        edges.append({\n",
    "            'src': f'cfRNA_{i}',\n",
    "            'dst': f'placenta_{j}',\n",
    "            'edge_type': 'cross_modality'\n",
    "        })\n",
    "\n",
    "# c) Optional: ancestry / gestational age edges\n",
    "# Example for ancestry: connect nodes of same ancestry within each dataset\n",
    "for df, prefix in zip([meta_cfRNA, meta_placenta], ['cfRNA','placenta']):\n",
    "    ancestry_map = df['ancestry'].to_dict()\n",
    "    nodes = list(df.index)\n",
    "    for i in range(len(nodes)):\n",
    "        for j in range(i+1, len(nodes)):\n",
    "            if ancestry_map[nodes[i]] == ancestry_map[nodes[j]]:\n",
    "                edges.append({\n",
    "                    'src': f'{prefix}_{i}',\n",
    "                    'dst': f'{prefix}_{j}',\n",
    "                    'edge_type': 'ancestry'\n",
    "                })\n",
    "\n",
    "# Convert to DataFrame and save\n",
    "edges_df = pd.DataFrame(edges)\n",
    "edges_df.to_csv(os.path.join(data_dir, 'graph_edges.csv'), index=False)\n",
    "\n",
    "print(\"✅ Dataset built successfully: train.csv, test.csv, node_types.csv, graph_edges.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "661f84b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bffe3f0f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
